<a href="https://colab.research.google.com/github/babydoryinsearch/B4U_Workshop1/blob/main/Daria_Nishchenko_B4U_Workshop1_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**Lab 3 Assignment: Data Preparation with Pandas**
---
#**Лабораторна робота 3: Підготовка даних за допомогою Pandas**

This laboratory work is dedicated to TreesCount! 2015. It is a large-scale public science initiative where participants collected tree census data in New York City. The results were compiled in the 2015 census ([Link](https://data.cityofnewyork.us/Environment/2015-Street-Tree-Census-Tree-Data/pi5s-9p35)) and were used for planning local city development initiatives.

To begin, let's import this dataset using the commands below:

---

Данна лабораторна робота присвячена *TreesCount! 2015*. Це широкомасштабна громадська наукова ініціатива, учасники якої збирали дані перепису дерев у Нью-Йорку. Ці результати були зібрані в переписі 2015 року ([Link](https://data.cityofnewyork.us/Environment/2015-Street-Tree-Census-Tree-Data/pi5s-9p35)) та були використанні для планування місцевих ініціатив з розвитку міста.

Для початку імпортуйємо цей набір даних за допомогою команд нижче:






In [ ]:
import pandas as pd
lab1_df=pd.read_csv('https://data.cityofnewyork.us/api/views/5rq2-4hqu/rows.csv?accessType=DOWNLOAD')

Once we have obtained the new dataset, we can proceed to use head() to view only the beginning of the file.

---

Коли вже отримали новий набір даних, далі можемо почати з використанням head(), щоб переглянути лише початок файлу.






In [ ]:
lab1_df.head()

created_at  tree_id  block_id  \
0  08/27/2015   180683    348711   
1  09/03/2015   200540    315986   
2  09/05/2015   204026    218365   
3  09/05/2015   204337    217969   
4  08/30/2015   189565    223043   

                                        the_geom  tree_dbh  stump_diam  \
0  POINT (-73.84421521958048 40.723091773924274)         3           0   
1   POINT (-73.81867945834878 40.79411066708779)        21           0   
2  POINT (-73.93660770459083 40.717580740099116)         3           0   
3  POINT (-73.93445615919741 40.713537494833226)        10           0   
4   POINT (-73.97597938483258 40.66677775537875)        21           0   

  curb_loc status health                           spc_latin  ... st_assem  \
0   OnCurb  Alive   Fair                         Acer rubrum  ...       28   
1   OnCurb  Alive   Fair                   Quercus palustris  ...       27   
2   OnCurb  Alive   Good  Gleditsia triacanthos var. inermis  ...       50   
3   OnCurb  Alive   Good  Gleditsia triacanthos var. inermis  ...       53   
4   OnCurb  Alive   Good                     Tilia americana  ...       44   

  st_senate   nta            nta_name  boro_ct     state   Latitude  \
0        16  QN17        Forest Hills  4073900  New York  40.723092   
1        11  QN49          Whitestone  4097300  New York  40.794111   
2        18  BK90   East Williamsburg  3044900  New York  40.717581   
3        18  BK90   East Williamsburg  3044900  New York  40.713537   
4        21  BK37  Park Slope-Gowanus  3016500  New York  40.666778   

   longitude          x_sp           y_sp  
0 -73.844215  1.027431e+06  202756.768749  
1 -73.818679  1.034456e+06  228644.837379  
2 -73.936608  1.001823e+06  200716.891267  
3 -73.934456  1.002420e+06  199244.253136  
4 -73.975979  9.909138e+05  182202.425999  

[5 rows x 42 columns]


Next, we will begin creating a new reduced DataFrame that contains only the following columns: the scientific name of each tree ('spc_latin'), the common name ('spc_common'), the status ('status') which can have values "Alive", "Dead", or "Stump", the health ('health'), and the diameter ('tree_dbh' for intact trees and 'stump_diam' for stumps).

To start, we will create a Python list containing the column names of interest. Then, we will create a new DataFrame named 'lab1_df_reduced' that only contains these columns.

---

Далі розпочнемо створення нового зменшеного DataFrame, що містить лише наступні стовпці: з науковою назвою кожного дерева ('spc_latin'), загальною назвою ('spc_common'), стану - 'status' (значенням є «Живе», «Мертве» або «Пень»), здоров'я - 'health' і діаметр ('tree_dbh' для неушкоджених дерев та 'stump_diam' для пнів).

Для початку **створемо список Python, що містить назви стовпців, які вас цікавлять. Після створемо новий DataFrame під назвою «*lab1_df_reduced*», який містить лише ці стовпці.** 

In [ ]:
column_names = ['spc_latin', 'spc_common', 'status', 'health', 'tree_dbh', 'stump_diam']
lab1_df_reduced = lab1_df[column_names]
lab1_df_reduced

spc_latin           spc_common status health  \
0                              Acer rubrum            red maple  Alive   Fair   
1                        Quercus palustris              pin oak  Alive   Fair   
2       Gleditsia triacanthos var. inermis          honeylocust  Alive   Good   
3       Gleditsia triacanthos var. inermis          honeylocust  Alive   Good   
4                          Tilia americana      American linden  Alive   Good   
...                                    ...                  ...    ...    ...   
683783                   Quercus palustris              pin oak  Alive   Good   
683784                 Cladrastis kentukea  Kentucky yellowwood  Alive   Good   
683785                         Acer rubrum            red maple  Alive   Good   
683786                         Acer rubrum            red maple  Alive   Good   
683787                         Acer rubrum            red maple  Alive   Fair   

        tree_dbh  stump_diam  
0              3           0  
1             21           0  
2              3           0  
3             10           0  
4             21           0  
...          ...         ...  
683783        25           0  
683784         7           0  
683785        12           0  
683786         9           0  
683787        23           0  

[683788 rows x 6 columns]

Next, we will check our new DataFrame for the presence of NaN values that may hinder further analysis. To do this, we run the following code:

---

Далі, перевіримо наш новий DataFrame на наявність значень NaN, які можуть заважати подальшому аналізу. Для цього запускаємо код:

In [ ]:
nullValue_counts=lab1_df_reduced.isnull().sum()
nullValue_counts

spc_latin     31619
spc_common    31619
status            0
health        31616
tree_dbh          0
stump_diam        0
dtype: int64

Based on this, we can see that there are 31,619 NaN values in the 'health' column and columns related to names. Let's further investigate, trying to determine the cause.

We will select the rows that have NaN values from our reduced DataFrame using the following command.

---

Виходячи з цього, ми бачимо, що в стовпцях '*health*' і стовпцях пов'язаних з іменами є 31 619 значень NaN. Давайте досліджуватимемо далі, намагаючись з’ясувати причину.

Оберемо рядки, які мають NaN із нашого скороченого DataFrame за допомогою наступної команди.

In [ ]:
NaNsOnly=lab1_df_reduced.loc[(lab1_df_reduced['spc_latin'].isnull() ) | (lab1_df_reduced['spc_common'].isnull() ) | (lab1_df_reduced['health'].isnull() ) ]
NaNsOnly.head()

spc_latin spc_common status health  tree_dbh  stump_diam
61        NaN        NaN   Dead    NaN         2           0
307       NaN        NaN  Stump    NaN         0           9
370       NaN        NaN   Dead    NaN         4           0
494       NaN        NaN  Stump    NaN         0          17
556       NaN        NaN   Dead    NaN         2           0

Carefully reviewing the following rows, we notice that the status of the first five rows changed from "Alive" to "Dead" and "Stump". Based on this, it appears that there were issues with identifying the species and health status of the dead trees and stumps.

It is unlikely that the "status" column changed randomly, as it would imply that New York is predominantly filled with dead trees and stumps.

Let's run another command so that we can gather more information to determine if these problematic rows were the source of our NaN values.

---

Уважно переглядаємо наступні рядки, та звертаємо увагу, що статус перших п’яти рядків змінився з «*Живий*» на *«Мертвий»* і *«Пень»*. Виходячи з цього, виглядає так, наче були проблеми з визначенням виду та стану здоров’я мертвих дерев і пнів.

Малоймовірно, що стовпець «*статус*» змінився випадково, оскільки інакше це означало б, що Нью-Йорк в основному заповнений мертвими деревами та пнями.

Давайте запустимо ще одну команду, щоб ми могли отримати більше інформації, щоб дізнатися, чи проблемні рядки  були джерелом наших NaN.

In [ ]:
pd.value_counts(NaNsOnly['status'])

Stump    17654
Dead     13961
Alive        6
Name: status, dtype: int64

Based on the obtained results, it can be assumed that our reduced NaN DataFrame is mostly not random errors, but rather issues with identifying physically inaccessible trees.

Assuming this, we can now decide how we want to handle the NaN values by exploring different questions about the census. Let's calculate the average diameter for each tree, including dead and/or unidentified ones.

During the calculation of the mean value, we should avoid columns where the diameter is not specified, as including them in our calculations would lead to inaccuracies.

By executing the following command, we can see that there are 278 such rows.

---

На основі отриманих результатів, можна припустити, що наші зменшені NaN DataFrame здебільшого не є випадковими помилками, а натомість є проблеми з ідентифікацією фізично недоступних дерев.

Припускаючи це, тепер ми можемо вирішити, як ми хочемо поводитись з значеннями NaN, досліджуючи різні питання про перепис. Давайте отримаємо середній діаметр для кожного дерева, включаючи мертві та/або неідентифіковані.

Під час обчислення середнього значення слід уникати стовпців, для яких не вказано діаметр, оскільки включення їх у наші розрахунки призведе до їх некорректності. 

Виконуючи наступну команду, ми бачимо, що є 278 таких рядків.

In [ ]:
lab1_df_reduced.loc[(lab1_df_reduced['tree_dbh'] == 0) & (lab1_df_reduced['stump_diam'] == 0)]

spc_latin      spc_common status health  tree_dbh  stump_diam
644                   NaN             NaN   Dead    NaN         0           0
3647                  NaN             NaN   Dead    NaN         0           0
5291    Cotinus coggygria       smoketree  Alive   Good         0           0
6029                  NaN             NaN   Dead    NaN         0           0
6114                  NaN             NaN   Dead    NaN         0           0
...                   ...             ...    ...    ...       ...         ...
677527                NaN             NaN   Dead    NaN         0           0
677676                NaN             NaN   Dead    NaN         0           0
677698                NaN             NaN   Dead    NaN         0           0
680500      Acer palmatum  Japanese maple  Alive   Good         0           0
682336                NaN             NaN   Dead    NaN         0           0

[278 rows x 6 columns]

Thus, let's filter out these rows and save a new DataFrame named 'lab1_df_reduced_zerosFiltered'. By using the command provided below, we can confirm that this filtered DataFrame is indeed 278 rows shorter than the previous one.

---

Таким чином, **відфільтруємо ці рядки та зберігаємо новий DataFrame під назвою '_lab1_df_reduced_zerosFiltered_'.** Використовуючи наведену нижче команду, підтверджуємо, що цей відфільтрований DataFrame точно на 278 рядків коротший, ніж попередній.

In [ ]:
lab1_df_reduced_zerosFiltered = lab1_df_reduced.loc[(lab1_df_reduced['tree_dbh'] != 0) | (lab1_df_reduced['stump_diam'] != 0)]

print(len(lab1_df_reduced) - len(lab1_df_reduced_zerosFiltered))

278


Next, let's perform the following steps:

1) Group the rows of the filtered DataFrame by the columns 'status', 'spc_latin', 'spc_common', and 'health'.

2) Add 'dropna=False' as an argument to groupby().

3) Use mean() to get the average tree diameter for each group.

4) Save the results as a DataFrame named 'lab1_df_reduced_zerosFiltered_grouped'.

Since we have made several modifications to our original dataset and will continue preparing the DataFrame for further analysis, let's also reset its index column.

---

Далі виконаємо наступні дії:

**1) Групуємо рядки відфільтрованого DataFrame за стовпцями '_status_', '_spc_latin_', '_spc_common_' і '_health_'.**

**2) Додаємо 'dropna=False' як аргумент до groupby().**

**3) Використовуємо mean(), щоб отримати середній діаметр дерева для кожної групи.**

**4) Зберігаємо результати як DataFrame під назвою 'lab1_df_reduced_zerosFiltered_grouped'**

Оскільки ми внесли багато змін до нашого оригінального набору даних, і будемо продовжувати, щоб підготувати DataFrame до цього, також скинемо його стовпці індексу.

In [ ]:
lab1_df_reduced_zerosFiltered_grouped = lab1_df_reduced_zerosFiltered.groupby(['status', 'spc_latin', 'spc_common', 'health'], dropna=False)[['tree_dbh', 'stump_diam']].mean().reset_index()
lab1_df_reduced_zerosFiltered_grouped

status                           spc_latin     spc_common health  \
0    Alive                                Acer          maple   Fair   
1    Alive                                Acer          maple   Good   
2    Alive                                Acer          maple   Poor   
3    Alive                   Acer buergerianum  trident maple   Fair   
4    Alive                   Acer buergerianum  trident maple   Good   
..     ...                                 ...            ...    ...   
393  Alive                                 NaN            NaN   Good   
394  Alive                                 NaN            NaN   Poor   
395   Dead  Gleditsia triacanthos var. inermis    honeylocust    NaN   
396   Dead                                 NaN            NaN    NaN   
397  Stump                                 NaN            NaN    NaN   

      tree_dbh  stump_diam  
0    11.903991    0.000000  
1    11.778865    0.000000  
2    10.509434    0.000000  
3     9.750000    0.000000  
4     7.724138    0.000000  
..         ...         ...  
393  13.000000    0.000000  
394  11.000000    0.000000  
395  12.000000    0.000000  
396   5.492017    0.000000  
397   0.000000   16.750481  

[398 rows x 6 columns]

The next step is to create a new column named "diameter" that combines the columns "tree_dbh" and "stump_diam". Since all rows have a non-zero value in one column but not in both, we can safely sum the two columns.

Calling the name of the DataFrame again will display the following:

---

Наступним кроком  **створемо новий стовпець під назвою «*diameter*», який поєднує стовпці «tree_dbh» і «stump_diam».** Оскільки всі рядки мають ненульове значення в одному стовпці, але не в обох, ми можемо підсумовати два стовпці.

Повторюючи виклик імені DataFrame виводить наступне:

In [ ]:
lab1_df_reduced_zerosFiltered_grouped['diameter'] = lab1_df_reduced_zerosFiltered_grouped[['tree_dbh', 'stump_diam']].fillna(0).sum(axis=1)
lab1_df_reduced_zerosFiltered_grouped

status                           spc_latin     spc_common health  \
0    Alive                                Acer          maple   Fair   
1    Alive                                Acer          maple   Good   
2    Alive                                Acer          maple   Poor   
3    Alive                   Acer buergerianum  trident maple   Fair   
4    Alive                   Acer buergerianum  trident maple   Good   
..     ...                                 ...            ...    ...   
393  Alive                                 NaN            NaN   Good   
394  Alive                                 NaN            NaN   Poor   
395   Dead  Gleditsia triacanthos var. inermis    honeylocust    NaN   
396   Dead                                 NaN            NaN    NaN   
397  Stump                                 NaN            NaN    NaN   

      tree_dbh  stump_diam   diameter  
0    11.903991    0.000000  11.903991  
1    11.778865    0.000000  11.778865  
2    10.509434    0.000000  10.509434  
3     9.750000    0.000000   9.750000  
4     7.724138    0.000000   7.724138  
..         ...         ...        ...  
393  13.000000    0.000000  13.000000  
394  11.000000    0.000000  11.000000  
395  12.000000    0.000000  12.000000  
396   5.492017    0.000000   5.492017  
397   0.000000   16.750481  16.750481  

[398 rows x 7 columns]

After that, we will execute the following command to remove the columns 'tree_dbh' and 'stump_diam':

---

Після цього **виконцємо таку команду, щоб видалити стовпці '*tree_dbh*' і '*stump_diam*':**

In [ ]:
lab1_df_reduced_zerosFiltered_grouped.drop(columns=['tree_dbh', 'stump_diam'], inplace=True)
lab1_df_reduced_zerosFiltered_grouped

status                           spc_latin     spc_common health  \
0    Alive                                Acer          maple   Fair   
1    Alive                                Acer          maple   Good   
2    Alive                                Acer          maple   Poor   
3    Alive                   Acer buergerianum  trident maple   Fair   
4    Alive                   Acer buergerianum  trident maple   Good   
..     ...                                 ...            ...    ...   
393  Alive                                 NaN            NaN   Good   
394  Alive                                 NaN            NaN   Poor   
395   Dead  Gleditsia triacanthos var. inermis    honeylocust    NaN   
396   Dead                                 NaN            NaN    NaN   
397  Stump                                 NaN            NaN    NaN   

      diameter  
0    11.903991  
1    11.778865  
2    10.509434  
3     9.750000  
4     7.724138  
..         ...  
393  13.000000  
394  11.000000  
395  12.000000  
396   5.492017  
397  16.750481  

[398 rows x 5 columns]

Our final modification to this DataFrame will involve adding a count for each row. To start, we assign a count of 1 to each row in the reduced but ungrouped DataFrame ('lab1_df_reduced') using the code provided below:

---

Наша остання модифікація цього DataFrame полягатиме в додаванні рахунку для кожного рядку. Для початку **призначаємо кожному рядку у скороченому, але не згрупованому DataFrame ('*lab1_df_reduced*') лічильник 1, використовуючи наведений нижче код:**

In [ ]:
lab1_df_reduced_counts=lab1_df_reduced.assign(count = 1)
lab1_df_reduced_counts.head()

spc_latin       spc_common status health  \
0                         Acer rubrum        red maple  Alive   Fair   
1                   Quercus palustris          pin oak  Alive   Fair   
2  Gleditsia triacanthos var. inermis      honeylocust  Alive   Good   
3  Gleditsia triacanthos var. inermis      honeylocust  Alive   Good   
4                     Tilia americana  American linden  Alive   Good   

   tree_dbh  stump_diam  count  
0         3           0      1  
1        21           0      1  
2         3           0      1  
3        10           0      1  
4        21           0      1

Next, we perform the following steps on 'lab1_df_reduced_counts':

1) Group the rows of the filtered DataFrame by the columns 'status', 'spc_latin', 'spc_common', and 'health'.

2) Add 'dropna=False' as an argument to groupby().

3) Use sum() to add the count for each group.

4) Save this as a DataFrame named 'lab1_df_reduced_counts_grouped'.

---

Далі виконуємо наступні кроки з *'lab1_df_reduced_counts'*:

**1) Групуємо рядки відфільтрованого DataFrame за стовпцями '_status_', '_spc_latin_', '_spc_common_' і '_health_'.**

**2) Додаємо 'dropna=False' як аргумент до groupby().**

**3) Використовуємо sum(), щоб додати кількість для кожної групи.**

**4) Зберігаємо це як DataFrame під назвою 'lab1_df_reduced_counts_grouped'**

In [ ]:
lab1_df_reduced_counts_grouped = lab1_df_reduced_counts.groupby(['status', 'spc_latin', 'spc_common', 'health'], dropna=False)
lab1_df_reduced_counts_grouped = lab1_df_reduced_counts_grouped.sum()
lab1_df_reduced_counts_grouped

tree_dbh  \
status spc_latin                          spc_common    health             
Alive  Acer                               maple         Fair       20582   
                                                        Good       55396   
                                                        Poor        6684   
       Acer buergerianum                  trident maple Fair         195   
                                                        Good         672   
...                                                                  ...   
       NaN                                NaN           Good          52   
                                                        Poor          11   
Dead   Gleditsia triacanthos var. inermis honeylocust   NaN           12   
       NaN                                NaN           NaN        76361   
Stump  NaN                                NaN           NaN            0   

                                                                stump_diam  \
status spc_latin                          spc_common    health               
Alive  Acer                               maple         Fair             0   
                                                        Good             0   
                                                        Poor             0   
       Acer buergerianum                  trident maple Fair             0   
                                                        Good             0   
...                                                                    ...   
       NaN                                NaN           Good             0   
                                                        Poor             0   
Dead   Gleditsia triacanthos var. inermis honeylocust   NaN              0   
       NaN                                NaN           NaN              0   
Stump  NaN                                NaN           NaN         295713   

                                                                count  
status spc_latin                          spc_common    health         
Alive  Acer                               maple         Fair     1733  
                                                        Good     4708  
                                                        Poor      639  
       Acer buergerianum                  trident maple Fair       20  
                                                        Good       87  
...                                                               ...  
       NaN                                NaN           Good        4  
                                                        Poor        1  
Dead   Gleditsia triacanthos var. inermis honeylocust   NaN         1  
       NaN                                NaN           NaN     13960  
Stump  NaN                                NaN           NaN     17654  

[398 rows x 3 columns]

The next step, to simplify the dataset, is to remove unnecessary columns and reset the index of this DataFrame using the following code:

---

Наступним кроком, щоб спростити набір даних, **видаляємо непотрібні стовпці та скидаємо індекс цього DataFrame за допомогою цього коду:** 

In [ ]:
lab1_df_reduced_counts_grouped.drop(columns=['tree_dbh','stump_diam'], inplace=True)
lab1_df_reduced_counts_grouped=lab1_df_reduced_counts_grouped.reset_index()
lab1_df_reduced_counts_grouped.head()

status          spc_latin     spc_common health  count
0  Alive               Acer          maple   Fair   1733
1  Alive               Acer          maple   Good   4708
2  Alive               Acer          maple   Poor    639
3  Alive  Acer buergerianum  trident maple   Fair     20
4  Alive  Acer buergerianum  trident maple   Good     87

As the final modification of our data, we create the final DataFrame named 'lab1_df_final' by merging the "left" (how='left') two DataFrames, 'lab1_df_reduced_counts_grouped' and 'lab1_df_reduced_zerosFiltered_grouped', in that order from left to right.*

After calling its name, we obtain the final DataFrame:

---

В якості останньої модифікації наших даних **створюємо остаточний Dataframe з назвою '*lab1_df_final*', об'єднавши "ліворуч"* (_how='left'_) два DataFrame-и '*lab1_df_reduced_counts_grouped*' і '*lab1_df_reduced_zerosFiltered_grouped*', у цьому порядку зліва направо.**

Після виклику його назви кінцевий DataFrame отримуємо:

In [ ]:
lab1_df_final = lab1_df_reduced_counts_grouped.merge(lab1_df_reduced_zerosFiltered_grouped, how='left')
lab1_df_final

status                           spc_latin     spc_common health  count  \
0    Alive                                Acer          maple   Fair   1733   
1    Alive                                Acer          maple   Good   4708   
2    Alive                                Acer          maple   Poor    639   
3    Alive                   Acer buergerianum  trident maple   Fair     20   
4    Alive                   Acer buergerianum  trident maple   Good     87   
..     ...                                 ...            ...    ...    ...   
393  Alive                                 NaN            NaN   Good      4   
394  Alive                                 NaN            NaN   Poor      1   
395   Dead  Gleditsia triacanthos var. inermis    honeylocust    NaN      1   
396   Dead                                 NaN            NaN    NaN  13960   
397  Stump                                 NaN            NaN    NaN  17654   

      diameter  
0    11.903991  
1    11.778865  
2    10.509434  
3     9.750000  
4     7.724138  
..         ...  
393  13.000000  
394  11.000000  
395  12.000000  
396   5.492017  
397  16.750481  

[398 rows x 6 columns]


To improve its appearance in the notebook, we execute the following:

---

Також покращуємо його вигляд у блокноті, виконавши наступне:

In [ ]:
lab1_df_final=lab1_df_final.set_index(['status','spc_latin','spc_common','health'])
lab1_df_final

count  \
status spc_latin                          spc_common    health          
Alive  Acer                               maple         Fair     1733   
                                                        Good     4708   
                                                        Poor      639   
       Acer buergerianum                  trident maple Fair       20   
                                                        Good       87   
...                                                               ...   
       NaN                                NaN           Good        4   
                                                        Poor        1   
Dead   Gleditsia triacanthos var. inermis honeylocust   NaN         1   
       NaN                                NaN           NaN     13960   
Stump  NaN                                NaN           NaN     17654   

                                                                 diameter  
status spc_latin                          spc_common    health             
Alive  Acer                               maple         Fair    11.903991  
                                                        Good    11.778865  
                                                        Poor    10.509434  
       Acer buergerianum                  trident maple Fair     9.750000  
                                                        Good     7.724138  
...                                                                   ...  
       NaN                                NaN           Good    13.000000  
                                                        Poor    11.000000  
Dead   Gleditsia triacanthos var. inermis honeylocust   NaN     12.000000  
       NaN                                NaN           NaN      5.492017  
Stump  NaN                                NaN           NaN     16.750481  

[398 rows x 2 columns]

Finally, we save the final DataFrame as a .csv file and submit it along with our Google Colab notebook for evaluation.

---

Нарешті, **зберігаємо остаточний DataFrame як файл .csv і надсилаємо його разом зі своїм блокнотом Google Colab для оцінювання.**

In [ ]:
lab1_df_final.to_csv('Nishchenko_lab1_df_final.csv')